File: crypto_agent/tools/compact.py
from .registry import register_tool


@register_tool(
    name="compact",
    description=(
        "Compress conversation history to free up context space.\n"
        "Use when the conversation is getting long and you need more room.\n"
        "Saves full transcript to .transcripts/ before compressing."
    ),
    schema={
        "type": "object",
        "properties": {},
    },
)
async def handle_compact(agent, **_) -> str:
    from ..context import auto_compact, estimate_tokens

    before = estimate_tokens(agent.messages)
    agent.messages = auto_compact(
        agent.messages,
        agent.client,
        agent.provider,
        force=True,
    )
    after = estimate_tokens(agent.messages)
    return f"Context compacted: {before:,} → {after:,} tokens (saved {before - after:,})"

File: src/tools/compact.ts
import { registerTool } from "./registry.js";

registerTool(
  "compact",
  "Compress conversation history to free up context space.\nUse when the conversation is getting long and you need more room.\nSaves full transcript to .transcripts/ before compressing.",
  { type: "object", properties: {} },
  async ({ agent, sessionId }) => {
    const { autoCompact, estimateTokens } = await import("../context.js");
    const session = agent.sessions.get(sessionId);
    const before = estimateTokens(session.messages);
    session.messages = await autoCompact(session.messages, agent.client, agent.provider, null, { force: true });
    const after = estimateTokens(session.messages);
    return `Context compacted: ${before.toLocaleString()} \u2192 ${after.toLocaleString()} tokens (saved ${(before - after).toLocaleString()})`;
  },
);
